---
title: "BIODEP DGE"
output: html_notebook
---

Loading required libraries


```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(reshape2)
library(tibble)
library(edgeR)
library(xCell)
```

-------------------
Not cell-corrected
-------------------

Create design matrix for DGE analysis

```{r}
de_catg <- final_metadata %>% mutate(centre = as.factor(as.character(Centre)), plate = as.factor(as.character(Plate.ID)), sex = as.factor(as.character(Sex)), disorder = as.factor(if_else(study_arm == 1, "CNT", "MDD"))) %>% dplyr::select(centre, disorder, plate, sex, Age)

design_pooled <- model.matrix(~disorder + centre + plate + sex + Age, data = de_catg)

de_catg_female <- final_metadata_female %>% mutate(centre = as.factor(as.character(Centre)), plate = as.factor(as.character(Plate.ID)), disorder = as.factor(if_else(study_arm == 1, "CNT", "MDD"))) %>% dplyr::select(centre, disorder, plate, Age)

design_female <- model.matrix(~ disorder + centre + plate + Age, data = de_catg_female)

de_catg_male <- final_metadata_male %>% mutate(centre = as.factor(as.character(Centre)), plate = as.factor(as.character(Plate.ID)), disorder = as.factor(if_else(study_arm == 1, "CNT", "MDD"))) %>% dplyr::select(centre, disorder, plate, Age)

design_male <- model.matrix(~ disorder + centre + plate + Age, data = de_catg_male)
```

Conduct DGE analysis using limma-voom

```{r}
voom_pooled <- voomWithQualityWeights(data_filtered, design = design_pooled, plot = T) 
voom_fit <- lmFit(voom_pooled, design = design_pooled)
voom_fit <- eBayes(voom_fit)
biodep_tt_ncc_pooled <- topTable(voom_fit, coef = "disorderMDD", number = nrow(data_filtered))
results <- decideTests(voom_fit)
summary(results)


voom_m <- voomWithQualityWeights(data_filtered_male, design = design_male, plot = T)
voom_fit_m <- lmFit(voom_m, design = design_male)
voom_fit_m <- eBayes(voom_fit_m)
biodep_tt_ncc_m <- topTable(voom_fit_m, coef = "disorderMDD", number = nrow(data_filtered_male))
results_m <- decideTests(voom_fit_m)
summary(results_m)


voom_f <- voomWithQualityWeights(data_filtered_female, design = design_female, plot = T)
voom_fit_f <- lmFit(voom_f, design = design_female)
voom_fit_f <- eBayes(voom_fit_f)
biodep_tt_ncc_f <- topTable(voom_fit_f, coef = "disorderMDD", number = nrow(data_filtered_female))
results_f <- decideTests(voom_fit_f)
summary(results_f)
```

-----------------------
Cell-corrected analysis
-----------------------

Load cell count data

```{r}
cell_counts <- read.delim("Apollo_qm_combi.csv", sep = ",", header = T, stringsAsFactors = F, row.names = 1) %>% dplyr::select(name, macs_flow_centre, macs_b_of_total:macs_nk_cd56hi_of_total) 
```

Retain datasets with cell count information

```{r}
final_metadata <- final_metadata %>% merge(., cell_counts, by.x = "Correct_Subject.ID", by.y = "name") %>% mutate(disorder = if_else(study_arm == 1, "CNT", "MDD"))

pdf("cellcount_plot.pdf", width = 8, height = 7)
boxplot(final_metadata[21:29]*100, ylab = "Percentage of PBMCs", las = 2, par(mar = c(15,5,0.5,0.5)))
dev.off()
```

Impute missing values
#code credit M-E Lynall

```{r}
cell_names <- c("macs_b_of_total", "macs_cd4_of_total", "macs_cd8_of_total", "macs_mono_class_of_total", "macs_mono_int_of_total", "macs_mono_nonclass_of_total", "macs_nk_cd56hi_of_total","macs_nkt_of_total","macs_nk_cd16hi_of_total")


#Turn cell count proportions into percentages to make LFC more meaningful
final_metadata[cell_names] <- lapply(final_metadata[cell_names], function(x) 100 * x)
```


```{r}
library(table1)

rndr <- function(x, name, ...) {
    if (length(x) == 0) {
        y <- final_metadata[[name]]
        s <- rep("", length(render.default(x=y, name=name, ...)))
        if (is.numeric(y)) {
            p <- wilcox.test(y ~ final_metadata$disorder)$p.value
        } else {
            p <- chisq.test(table(y, droplevels(final_metadata$disorder)))$p.value
        }
        s[2] <- sub("<", "&lt;", format.pval(p, digits=3, eps=0.001))
        s
    } else {
        render.default(x=x, name=name, ...)
    }
}

tmp <- final_metadata
tmp$disorder <- factor(as.character(final_metadata$disorder), levels=c("CNT","MDD","P-value"))

rndr.strat <- function(label, n, ...) {
    ifelse(n==0, label, render.strat.default(label, n, ...))
}

table1(~ macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_int_of_total + macs_mono_nonclass_of_total + macs_nk_cd56hi_of_total + macs_nkt_of_total + macs_nk_cd16hi_of_total | disorder, data=tmp, overall="Total", droplevels=F, render=rndr, render.strat=rndr.strat)
#image opened in html and screenshot taken
```

```{r}
# First look at missing values problem
library(visdat)
library(naniar)
library(mclust)
library(mix)


vis_miss(final_metadata[,cell_names])
ggsave(filename = "miss_counts_to_impute.pdf", width=6, height=6)

#all entries have a macs so used downstream as such

# Careful - MUST make factor if want treated as categorical
final_metadata$disorder <- factor(final_metadata$disorder) 
final_metadata$macs_flow_centre <- factor(final_metadata$macs_flow_centre)
final_metadata$Sex <- factor(final_metadata$Sex)

```


```{r}

set.seed(111)
print("Imputing using mix package")

input <- final_metadata[, c("Age", "Sex", "disorder", "macs_flow_centre", cell_names)]
str(input)
nimp <- sum(is.na(input))
nimp2 <- nrow(input)*ncol(input)
print(paste("Number of elements imputed in total = ",nimp," out of ",nimp2," total elements as follows: (",signif(nimp/nimp2*100,2)," %)",sep=""))
print(sapply(input,function(x) sum(is.na(x))))
print("Now impute")
tmp <- imputeData(input,seed=111,verbose=TRUE)
tmp <-tmp %>% as.data.frame()
tmp$Correct_Subject.ID <- final_metadata$Correct_Subject.ID
print(head(tmp))
# Drop unimputed and join imputed
imputed_metadata <- final_metadata %>% select(-one_of(cell_names)) %>% full_join(tmp[,c("Correct_Subject.ID",cell_names)],by="Correct_Subject.ID")

rm(tmp, input, nimp, nimp2)
```

Combine cells with low individual counts

```{r}
imputed_metadata <- imputed_metadata %>% mutate(macs_nk_of_total = macs_nk_cd16hi_of_total + macs_nk_cd56hi_of_total, macs_mono_cd16hi_of_total = macs_mono_int_of_total + macs_mono_nonclass_of_total) 
```


Create design matrix for DGE analysis

```{r}
imputed_metadata$Centre <- as.factor(imputed_metadata$Centre)
imputed_metadata$Plate.ID <- as.factor(imputed_metadata$Plate.ID)
imputed_metadata$Sex <- as.factor(imputed_metadata$Sex)

design_pooled_cc <- model.matrix(~ disorder + Centre + Plate.ID + Sex + Age + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total + macs_nk_of_total, data = imputed_metadata)

imputed_metadata_fem <- imputed_metadata %>% filter(Sex == "Female")
imputed_metadata_fem$Centre <- as.factor(imputed_metadata_fem$Centre)
imputed_metadata_fem$Plate.ID <- as.factor(imputed_metadata_fem$Plate.ID)

design_female_cc <- model.matrix(~ disorder + Centre + Plate.ID + Age + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total + macs_nk_of_total, data = imputed_metadata_fem)

imputed_metadata_male <- imputed_metadata %>% filter(Sex == "Male")
imputed_metadata_male$Centre <- as.factor(imputed_metadata_male$Centre)
imputed_metadata_male$Plate.ID <- as.factor(imputed_metadata_male$Plate.ID)

design_male_cc <- model.matrix(~ disorder + Centre + Plate.ID + Age + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total +  macs_nk_of_total, data = imputed_metadata_male)
```

Re-filter and re-normalise count data subset with cell count information

```{r}
counts_samples_cc <- counts_samples %>% dplyr::select(one_of(dput(as.character(imputed_metadata$sample_id_match))))
data_filtered_cc<- DGEList(counts = counts_samples_cc, genes = gene_info,  group = imputed_metadata$disorder)
keep <- filterByExpr(data_filtered_cc)
data_filtered_cc <- data_filtered_cc[keep, , keep.lib.sizes=FALSE]
data_filtered_cc <- calcNormFactors(data_filtered_cc)
data_normalised_cc <- cpm(data_filtered_cc, log = T, prior.count = 0.5)


counts_samples_female_cc <- counts_samples %>% dplyr::select(one_of(dput(as.character(imputed_metadata_fem$sample_id_match))))
data_filtered_female_cc <- DGEList(counts = counts_samples_female_cc, genes = gene_info,  group = imputed_metadata_fem$disorder)
keep <- filterByExpr(data_filtered_female_cc)
data_filtered_female_cc <- data_filtered_female_cc[keep, , keep.lib.sizes=FALSE]
data_filtered_female_cc <- calcNormFactors(data_filtered_female_cc)
data_normalised_female_cc <- cpm(data_filtered_female_cc, log = T, prior.count = 0.5)


counts_samples_male_cc <- counts_samples %>% dplyr::select(one_of(dput(as.character(imputed_metadata_male$sample_id_match))))
data_filtered_male_cc <- DGEList(counts = counts_samples_male_cc, genes = gene_info,  group = imputed_metadata_male$disorder)
keep <- filterByExpr(data_filtered_male_cc)
data_filtered_male_cc <- data_filtered_male_cc[keep, , keep.lib.sizes=FALSE]
data_filtered_male_cc <- calcNormFactors(data_filtered_male_cc)
data_normalised_male_cc <- cpm(data_filtered_male_cc, log = T, prior.count = 0.5)
```


Conduct DGE analysis using limma-voom

```{r}
voom_pooled_cc <- voomWithQualityWeights(data_filtered_cc, design = design_pooled_cc, plot = T)
voom_fit_cc <- lmFit(voom_pooled_cc, design = design_pooled_cc)
voom_fit_cc <- eBayes(voom_fit_cc)
biodep_tt_cc_pooled <- topTable(voom_fit_cc, coef = "disorderMDD", number = nrow(data_filtered_cc))
results_cc <- decideTests(voom_fit_cc)
summary(results_cc)


voom_m_cc <- voomWithQualityWeights(data_filtered_male_cc, design = design_male_cc, plot = T)
voom_fit_m_cc <- lmFit(voom_m_cc, design = design_male_cc)
voom_fit_m_cc <- eBayes(voom_fit_m_cc)
biodep_tt_cc_m <- topTable(voom_fit_m_cc, coef = "disorderMDD", number = nrow(data_filtered_male_cc))
results_m_cc <- decideTests(voom_fit_m_cc)
summary(results_m_cc)


voom_f_cc <- voomWithQualityWeights(data_filtered_female_cc, design = design_female_cc, plot = T)
voom_fit_f_cc <- lmFit(voom_f_cc, design = design_female_cc)
voom_fit_f_cc <- eBayes(voom_fit_f_cc)
biodep_tt_cc_f <- topTable(voom_fit_f_cc, coef = "disorderMDD", number = nrow(data_filtered_female_cc))
results_f_cc <- decideTests(voom_fit_f_cc)
summary(results_f_cc)

```

---------------------------------
Data prep for downstream analysis
---------------------------------

```{r}
catgs <- imputed_metadata_fem %>% dplyr::select(sample_id_match, disorder, Centre, Plate.ID, Age, macs_b_of_total:macs_mono_class_of_total,  macs_mono_cd16hi_of_total, macs_nk_of_total) 
catgs[, c(2:4)] <- lapply(catgs[, c(2:4)], factor)

design = model.matrix(~ Centre + Plate.ID + Age +macs_b_of_total+macs_cd4_of_total+macs_cd8_of_total+macs_mono_class_of_total+macs_mono_cd16hi_of_total+macs_nk_of_total, data = catgs)
mm <- model.matrix(~disorder, data = catgs)
batch_correct_b_f <- removeBatchEffect(data_normalised_female_cc, covariates = design[,-1], design = mm)

catgs <- imputed_metadata_male %>% dplyr::select(sample_id_match, disorder, Centre, Plate.ID, Age, macs_b_of_total:macs_mono_class_of_total,  macs_mono_cd16hi_of_total, macs_nk_of_total) 
catgs[, c(2:4)] <- lapply(catgs[, c(2:4)], factor)

design = model.matrix(~ Centre + Plate.ID + Age +macs_b_of_total+macs_cd4_of_total+macs_cd8_of_total+macs_mono_class_of_total+macs_mono_cd16hi_of_total+macs_nk_of_total, data = catgs)
mm <- model.matrix(~disorder, data = catgs)
batch_correct_b_m <- removeBatchEffect(data_normalised_male_cc, covariates = design[,-1], design = mm)

catgs <- imputed_metadata %>% dplyr::select(sample_id_match, disorder, Centre, Plate.ID, Sex, Age, macs_b_of_total:macs_mono_class_of_total,  macs_mono_cd16hi_of_total, macs_nk_of_total) 
catgs[, c(2:4)] <- lapply(catgs[, c(2:4)], factor)

design = model.matrix(~ Centre + Plate.ID + Sex + Age +macs_b_of_total+macs_cd4_of_total+macs_cd8_of_total+macs_mono_class_of_total+macs_mono_cd16hi_of_total+macs_nk_of_total, data = catgs)
mm <- model.matrix(~disorder, data = catgs)
batch_correct_b_pooled <- removeBatchEffect(data_normalised_male_cc, covariates = design[,-1], design = mm)


metd_b <- imputed_metadata

save(metd_b, batch_correct_b_f, batch_correct_b_m, file = "../for_WGCNA/biodep.Rdata")
```


--------------------
Sensitivity analysis
--------------------


Prep count data for DGE analysis

```{r}
metd_sen <- imputed_metadata %>% filter(is.na(BMI.Calculated) == F)
metd_sen_f <- metd_sen %>% filter(Sex == "Female")
metd_sen_m <- metd_sen %>% filter(Sex == "Male")
  
data_filtered_sens <- data_filtered_cc$counts
data_filtered_sens_f <- data_filtered_female_cc$counts
data_filtered_sens_m <- data_filtered_male_cc$counts

data_filtered_sens <- data_filtered_sens %>% as.data.frame(.) %>% dplyr::select(one_of(dput(as.character(metd_sen$sample_id_match))))
data_filtered_sens_f <- data_filtered_sens_f %>% as.data.frame(.) %>% dplyr::select(one_of(dput(as.character(metd_sen_f$sample_id_match))))
data_filtered_sens_m <- data_filtered_sens_m %>% as.data.frame(.) %>% dplyr::select(one_of(dput(as.character(metd_sen_m$sample_id_match))))

data_filtered_sens <- DGEList(data_filtered_sens, group = metd_sen$disorder)
data_filtered_sens <- calcNormFactors(data_filtered_sens)
data_filtered_sens_f <- DGEList(data_filtered_sens_f, group = metd_sen_f$disorder)
data_filtered_sens_f <- calcNormFactors(data_filtered_sens_f)
data_filtered_sens_m <- DGEList(data_filtered_sens_m, group = metd_sen_m$disorder)
data_filtered_sens_m <- calcNormFactors(data_filtered_sens_m)
```

Create design matrix

```{r}
design_sens <- model.matrix(~ disorder + Centre + Plate.ID + Sex + Age + BMI.Calculated + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total + macs_nk_of_total, data = metd_sen)

design_sens_m <- model.matrix(~ disorder + Centre + Plate.ID + Age + BMI.Calculated + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total + macs_nk_of_total, data = metd_sen_m)

design_sens_f <- model.matrix(~ disorder + Centre + Plate.ID + Age + BMI.Calculated + macs_b_of_total + macs_cd4_of_total + macs_cd8_of_total + macs_mono_class_of_total + macs_mono_cd16hi_of_total + macs_nk_of_total, data = metd_sen_f)

```

```{r}
voom_pooled_sens <- voomWithQualityWeights(data_filtered_sens, design = design_sens, plot = T)
voom_fit_sens <- lmFit(voom_pooled_sens, design = design_sens)
voom_fit_sens <- eBayes(voom_fit_sens)
biodept_tt_sens_pooled <- topTable(voom_fit_sens, coef = "disorderMDD", number = nrow(data_filtered_sens))
results_sens <- decideTests(voom_fit_sens)
summary(results_sens)


voom_m_sens <- voomWithQualityWeights(data_filtered_sens_m, design = design_sens_m, plot = T)
voom_fit_m_sens <- lmFit(voom_m_sens, design = design_sens_m)
voom_fit_m_sens <- eBayes(voom_fit_m_sens)
biodep_tt_sens_m <- topTable(voom_fit_m_sens, coef = "disorderMDD", number = nrow(data_filtered_sens_m))
results_m_sens <- decideTests(voom_fit_m_sens)
summary(results_m_sens)


voom_f_sens <- voomWithQualityWeights(data_filtered_sens_f, design = design_sens_f, plot = T)
voom_fit_f_sens <- lmFit(voom_f_sens, design = design_sens_f)
voom_fit_f_sens <- eBayes(voom_fit_f_sens)
biodep_tt_sens_f <- topTable(voom_fit_f_sens, coef = "disorderMDD", number = nrow(data_filtered_sens_f))
results_f_sens <- decideTests(voom_fit_f_sens)
summary(results_f_sens)
```

--------------------------------------------------------------
xCell correction for comparison with original cell count data
--------------------------------------------------------------

Loading data for xCell analysis

```{r}
pooled_counts <- data_filtered[['counts']] %>% as.data.frame()
```

10,808 gene overlap check, at least 5000 recommended.

```{r}
load("../../../../ensembl_GRCh38_v84.Rdata")
att <- c("external_gene_name", "ensembl_gene_id", "chromosome_name")
gene_info <- getBM(attributes = att, mart = ensembl)

xcell_genes <- xCell.data$genes %>% as.data.frame() %>% setNames("gene_name")
pooled_genes <- rownames(pooled_counts) %>% as.data.frame() %>% setNames("ensembl_gene_id") %>% merge(., gene_info, by = "ensembl_gene_id")

venn.diagram(list(Pooled = pooled_genes$external_gene_name, xCell = xcell_genes$gene_name), filename = "overlap_10808_genes.tiff") 
```


```{r}
library(ensembldb)
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86
```


```{r}
gene_lengths <- lengthOf(edb, of = "gene")
gene_lengths <- as.data.frame(gene_lengths) %>% rownames_to_column("ensembl_gene_id")

gene_len_pooled <- pooled_genes %>% merge(., gene_lengths, by = "ensembl_gene_id", all.x = T)


chr_names <- c(as.character(paste(seq(1,22,1), sep = ",")),"X", "Y", "MT")
gene_len_pooled_2 <- gene_len_pooled %>% dplyr::filter(chromosome_name %in% chr_names) 
setdiff(gene_len_pooled$ensembl_gene_id, gene_len_pooled_2$ensembl_gene_id)#13

gene_len_pooled_3 <- gene_len_pooled_2 %>% dplyr::filter(is.na(gene_lengths) == F)

detach(package:EnsDb.Hsapiens.v86)
detach(package:ensembldb)
```


modify male and female count matrix - gene subset
```{r}
count_samples_sub <- pooled_counts %>% rownames_to_column("ensembl_gene_id") %>% merge(., gene_len_pooled_3, by = "ensembl_gene_id", sort = F) %>% mutate(gene_name = make.unique(external_gene_name, sep = "_")) 
gl_pooled <- count_samples_sub %>% dplyr::select(gene_name, gene_lengths)
count_samples_sub <- count_samples_sub %>% column_to_rownames("gene_name")

final_metadata_pooled <- imputed_metadata %>% mutate(Disorder = as.factor(disorder))
identical(final_metadata_pooled$sample_id_match, colnames(count_samples_sub))

count_samples_sub <- count_samples_sub %>% dplyr::select(one_of(dput(as.character(final_metadata_pooled$sample_id_match))))

```

```{r}
data_filtered_sub <- DGEList(counts = count_samples_sub, group = final_metadata_pooled$Disorder)

data_filtered_sub <- calcNormFactors(data_filtered_sub)
data_filtered_sub$samples
```

RPKM find

```{r}

RPKM_pooled <- rpkm(data_filtered_sub, gene.length = gl_pooled$gene_lengths)

```

TPM calculate (TMM normalised - at RPKM stage)

```{r}

TPM_pooled <- t( t(RPKM_pooled) / colSums(RPKM_pooled) ) * 1e6

```

xCell calculation

```{r}
cellchoice_broad <- c("B-cells", "CD4+ T-cells", "CD8+ T-cells", "Monocytes", "NK cells") 

set.seed(13005)
xcellout_broad <- xCellAnalysis(expr=TPM_pooled, genes=rownames(TPM_pooled),rnaseq=TRUE, cell.types.use=cellchoice_broad)

```

### xCell stats

#significance
```{r}
set.seed(13005)
# Significance testing for whether cell types are likely present in the mixture
tmp <- xCellSignifcanceBetaDist(xcellout_broad, rnaseq=TRUE)
# Only keep a cell type if there is evidence (p<0.05) for presence of the cell type in at least 30% samples (arbitrary)
toselect <- apply(tmp,1,function(x){sum(x<0.1)}>=64)
toselect %>% table 
xcellsmall_broad <- xcellout_broad[toselect,]
ypos_broad <- apply(xcellsmall_broad, 1, function(x){1.3*max(x)})

```

Final values to correct for

```{r}
cc_broad_pooled <- xcellsmall_broad %>% t() %>% as.data.frame() %>% rownames_to_column ("sample_id")
```

```{r}
orig_cc <- final_metadata %>% mutate(monocytes = macs_mono_class_of_total + macs_mono_int_of_total + macs_mono_nonclass_of_total) %>% dplyr::select(sample_id_match, macs_b_of_total:macs_cd8_of_total, monocytes, macs_nk_of_total)

identical(cc_broad_pooled$sample_id, orig_cc$sample_id_match)

```

Correlation

```{r}
cor_pooled <- rcorr(x = as.matrix(orig_cc[,-1]), y = as.matrix(cc_broad_pooled[,-1]), type = "spearman")

library(corrplot)

pdf("correlation_broads.pdf")
corrplot(corr = cor_pooled$r[1:5, 6:10], method = "square", title = "Pooled - BIODEP", p.mat = cor_pooled$P[1:5, 6:10], sig.level = 0.05, insig = "blank", mar = c(0,0,1,0), addCoef.col = "black")
dev.off()


```






